### PR TITLE
cloud/amazon: reuse s3 client sessions

### PR DESCRIPTION
**Pull request by dt**
This changes the s3 client wrapper to optionally cache the last S3 client opened based
on the arguments and then re-use that client for subsequent client opening to the same
bucket with the same session parameters.

This should avoid issues that arise from opening many sessions at once, like hammering
the EC2 metadata server which can then result in it rejecting some requests and the SDK
returning an error like 'NoCredentialProviders: no valid providers in chain'. This is a
particular concern when loading a long chain of incremental backups which, since #47158,
is done in parallel. That parallel loading can generate a flood of new S3 sessions at once
and quickly hit the metadata server's rate limit. With this session the first will establish
a new session, including loading credentials from the metadata server, and subsequent sessions
will reuse it.

Release note (bug fix): Operations like BACKUP can now reuse a previously created S3 client
session when operating on the same bucket, which can avoid NoCredentialProviders errors on
EC2 when interating with large incremental backups.

### CODE REVIEW

**Comment by dt** [comment]
Currently deployed on a node in my long-running test cluster (david-backup:6):
```
root@localhost:26257/defaultdb> backup table demo to 's3://cockroach-backup-tests/demo/2021/06/07-000000.00?AUTH=implicit';
        job_id       |  status   | fraction_completed | rows | index_entries | bytes
---------------------+-----------+--------------------+------+---------------+--------
  665367655012401158 | succeeded |                  1 |    0 |             0 |     0
(1 row)

Time: 1.986s total (execution 1.985s / network 0.000s)

root@localhost:26257/defaultdb> set cluster setting cloudstorage.s3.session_reuse.enabled = false;
SET CLUSTER SETTING

Time: 9ms total (execution 9ms / network 0ms)

root@localhost:26257/defaultdb> backup table demo to 's3://cockroach-backup-tests/demo/2021/06/07-000000.00?AUTH=implicit';
ERROR: failed to read backup from "s3://cockroach-backup-tests/demo/2021/06/07-000000.00/20210607/095500.00?AUTH=implicit": failed to get s3 object: NoCredentialProviders: no valid providers in chain
caused by: EnvAccessKeyNotFound: failed to find credentials in the environment.
SharedCredsLoad: failed to load profile, .
EC2RoleRequestError: failed to get backup-testing EC2 instance role credentials
caused by: RequestError: send request failed
caused by: Get "http://169.254.169.254/latest/meta-data/iam/security-credentials/backup-testing": read tcp 10.12.28.210:44572->169.254.169.254:80: read: connection reset by peer
```

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/storage/cloud/amazon/s3_storage.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-MbfXmIs3thFoFWi3n08:-MbfXmIs3thFoFWi3n09:bqeh9pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fe8e3b0cec70584f95b8a91a18360cd93b913e/pkg/storage/cloud/amazon/s3_storage.go#L77)):*
> ```Go
> 
> // s3Client wraps a client and uploader for the same session used to create the
> // client and can be stored for reuse by other requests using
> ```

incomplete senten....
___
*[pkg/storage/cloud/amazon/s3_storage.go, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-MbfTM285Vjw0wVnenZ9:-MbfTM285Vjw0wVnenZA:bcucsuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fe8e3b0cec70584f95b8a91a18360cd93b913e/pkg/storage/cloud/amazon/s3_storage.go#L237)):*
> ```Go
> 	}
> 
> 	reuse := reuseSession.Get(&args.Settings.SV)
> ```

I'm not sure I like this global setting; nor global s3clientCache...

I just don't see how we would want to have this setting `false` when we *know* that not using it could
cause issues.  On the other hand, we also know that using this setting could also cause problems e..g when we need to write out large files and the keys get rotated.

Another concern is that this change seems a bit large in scope, and I assume we would want to backport that.

I wonder if we would be better off if we let the call site indicate that  it's okay to reuse the session.
As things stand right now, we have `getBackupManifests` as the only place that has this problem.
What if we let the caller annotate context to indicate that it's okay to reuse the session in the storage?

```
type sessionReuseKey struct {
	
}
func WithSessionReuse(ctx context.Context) context.Context{
	return context.WithValue(ctx, sessionReuseKey{}, true)
}
func ReuseSession(ctx context.Context) bool {
	reuse, ok := ctx.Value(sessionReuseKey{}).(bool)
	return ok && reuse
}
```

I think the "blast" radius of such change would be pretty small ; you won't have to have global cache; and
we will not need to have global cluster setting either.

Thoughts?

**Comment by dt** [comment]
We've also seen `ExportRequest` hit these errors which makes sense as it can open, write and close many times very quickly e.g. if one node hosts lots of small ranges or ranges with small changes during in inc backup, and many requests can arrive and run in parallel. 

Whereas something like the backup coordinator could, in theory, open one `ExternalStorage` and then call methods on it several times, for example to open many manifests, and thus make use of a reused `*s3.S3` field within it across those requests (e.g. as we had in 20.1), because `ExportRequest`s are totally separate and each request has to make its own `ExternalStorage`, the global, process-level cache is the only option for reusing sessions across `ExportRequest`'s. Additionally, the most recent customers reporting issues with this is using explicit incremental backup locations, where they provide list of URIs to open, each of which is opened as a separate `ExternalStorage` (they could even be on totally separate clouds), so here too we want the process-level cache of s3 sessions so that these separate `ExternalStorage` instances can share a session.

So _when we want reuse_, the global cache is more or less forced. Now as far as how to decide when we want reuse, the global setting does indeed mean you need to pick: reused sessions or reload `credentials` file. I think I'm fine with this for now, particularly now that we've realized everyone who reported `NoCredentialProviders` errors wasn't actually using / seeing stale values from `credentials` files at all (since if they read a file at all, they would see `Invalid key` not `NoCredentialProviders`).

Ultimately I think the goal is _everyone_ uses session reuse and we just wrap the file provider with one that supports refreshing, but given our current understanding that ~nobody seeing credential rotation errors is using file-based credentials, that can happen at some later time.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Release note (bug fix): Operations like BACKUP can now reuse a previously created S3 client
> session when operating on the same bucket, which can avoid NoCredentialProviders errors on
> EC2 when interating with large incremental backups.

Can you update release note to be explicit that `... BACKUP now reuse previous created S3 client *by default*`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-:-Mbfu7BQ0jwdJBuArNZj:b-8hw25d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/storage/cloud/amazon/s3_storage.go, line 245 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-Mbfpn7MBfJys3EkOTZC:-Mbfpn7MBfJys3EkOTZD:bm7yze3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c12c8865444e971fa786d302b67554795d6495c/pkg/storage/cloud/amazon/s3_storage.go#L245)):*
> ```Go
> 		cacheKey.conf.Prefix = ""
> 		cacheKey.conf.ServerEncMode = ""
> 		cacheKey.conf.ServerKMSID = ""
> ```

Few comments re cache:

Perhaps it's a bit more work, but would introduce

```
type clientCache interface {
  getClient(context.Context, bucket string, opts session.Options) ((*s3Client, error) 
  put(s3Client)...
}
```

Basically, two implementations (`alwaysCreateClient` and `cachingClientCache`), and a helper
method `getClientCache()` to return appropriate cache based setting value.

I think pulling "newClient" into interface like this would simpify this function too.  you just call 
getClient()/putClient on the cache object.

Secondly, I wonder why we use conf as the cache key.  Shouldn't we use session.options instead?

Finally, would it be too difficult to support more than one cache item?
I don't think we even need to worry too much about the "efficiency" here.  The cache could simply keep
array of s3clients (with their applicable config), and just iterate until we find a match.

**Comment by dt** [comment]
I didn't mention `*by default*` in the release note since I didn't even mention that it is optional, without which "by default" seems duplicative? 

I did a little refactoring: I didn't go as far as separate client factories and swapping implementations, but I did change the factory to be a function instead of a method to make its inputs easy to audit and more importantly made it use as input the same struct we use as a the cache key. Not we don't use session.Options since it is very pointer-y and thus hashes/compares poorly, but with this refactor a session.Options is a pure function of our strings-only struct `s3ClientOption`, so we use the latter and derive the former just in time.

I mentioned offline, but I think we can make the cache multi-element in a very contained follow-up, that has more limited scope. Until then, even just 1-elem cache should still be a win for the bursts of requests we get within a single backup so I don't really want to add scope to this diff.

**Review by miretskiy — APPROVED** [review]
Ack.  I just thought it's important to call out that that the default is to reuse.
I'm okay with your approach -- I was just trying to suggest something that I think improved code readability -- and your approach does that as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-:-MbgLDoT72NtNEDOModb:bm7chhg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/storage/cloud/amazon/s3_storage.go, line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-MbgLUfhESjOdxiu5I-w:-MbgLUfhESjOdxiu5I-x:b-wbuu1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/97dec9c3549eaca68cedf223db655013eef1a26e/pkg/storage/cloud/amazon/s3_storage.go#L90)):*
> ```Go
> 
> // s3ClientConfig is the immutable config used to initialize an s3 session.
> // It contains values copied from cooresponding fields in ExternalStorage_S3
> ```

s/coor/cor/
___
*[pkg/storage/cloud/amazon/s3_storage.go, line 266 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-MbgLh-6AxyayxoMxcR0:-MbgLh-6AxyayxoMxcR1:b-r6ajxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/97dec9c3549eaca68cedf223db655013eef1a26e/pkg/storage/cloud/amazon/s3_storage.go#L266)):*
> ```Go
> 	}
> 
> 	client, _, err := newClient(ctx, s.opts, s.settings)
> ```

I think we should create client outside of cache lock.  It's okay if you wind up overriding cached client in case of a race; but it's probably less okay to keep cache locked for potentially long period while trying to create newClient (which does involve external IO)

**Review by miretskiy — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/storage/cloud/amazon/s3_storage.go, line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-MbgLUfhESjOdxiu5I-w:-MbgMOk88xBoCpmEsrBV:b-7rz21o) ([raw file](https://github.com/cockroachdb/cockroach/blob/97dec9c3549eaca68cedf223db655013eef1a26e/pkg/storage/cloud/amazon/s3_storage.go#L90)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

s/coor/cor/
</blockquote></details>

s/coor/corr/

**Review by pbardea — APPROVED** [review]
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

**Comment by dt** [comment]
Heh, I originally had create-outside-lock, with the same thinking -- use the cached one if it is ready, but don't wait if it isn't and risk making a wasted duplicate client in the race case -- but I changed it in the latest revision: I think we actually _want_ additional would-be creators to be forced wait, since it isn't _just_ about avoiding "wasted" duplicate clients but also that we want to rate limit new client creations regardless (the whole reason we're making this change). In that context, holding the lock the whole time makes sense I think.

**Review by miretskiy — COMMENTED** [review]
I'm worried about the case when we have misconfigured or otherwise slow requests... 

Two things: I would add a comment why we think it's okay to hold the lock; 2.  We should probably create some sort of a span to trace potentially slow requests.... Or, alternatively, since the intention is for the client creation to be restricted to only 1 at a time, maybe use `ConcurrentRequestLimiter` which does the required throttling and tracing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66172#-:-MbgSYLpBzIkPIfdhAQH:b-y3ho40)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

**Comment by dt** [comment]
TFTRs!

bors r+
